[i18n] Remove @cog_i18n decorator

Tobotimus opened the pull request:
Continues on from #2206.

These are just general design improvements and some cleanups for i18n, possible now through the `commands.Cog` class and by extension, the `commands.CogMeta` metaclass.
- `Translator` no longer accepts a name as its first argument, as it's unused and unnecessary
- Removes @cog_i18n decorator in favour of passing the translator as an argument to `commands.Cog.__init_subclass__`
- Assigns `__translator` object to cog classes, and adds new method `commands.CogMeta.translate()` which is used by both commands and cogs to translate help strings
- Removes `commands.Command.translator` attribute

Tobotimus committed:
Refactor commands.Cog, create commands.CogMeta

This does quite a few things:
- Moved `commands.Cog` into new module `redbot.core.commands.cog`.
- `commands.Cog` inherits from `commands.CogMeta`, which in turn inherits from what `commands.Cog` previously inherited from.
- `requires` attribute is now attached to cog classes rather than cog instances, as is more consistent with commands.
- Rules are now cleared from the cache of cogs/commands when they are unloaded (previously this could have resulted in issues for commands if the rules changed whilst the command was unloaded)
- Fixed permissions hooks being cleaned up incorrectly in `Red.remove_cog` (this is something which slipped through from #2149, I meant to come back and fix this before that PR was merged)
- Cogs can now be disabled in particular guilds, in the same way commands can (I purposefully avoided allowing a global disable since users should really unload the cog instead)
- Cog classes can now be decorated with checks, and such checks will apply to all top-level commands

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Tobotimus committed:
[i18n] Refactor @cog_i18n and Translator

These are just general design improvements and some cleanups for i18n, possible now through the `commands.Cog` class and by extension, the `commands.CogMeta` metaclass.
- `Translator` no longer accepts a name as its first argument, as it's unused and unnecessary
- Removes @cog_i18n decorator in favour of passing the translator as an argument to `commands.Cog.__init_subclass__`
- Assigns `__translator` object to cog classes, and adds new method `commands.CogMeta.translate()` which should be used by both commands and cogs to translate help strings
- Removes `commands.Command.translator` attribute

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

mikeshardmind commented on redbot/core/commands/cog.py:
```diff
@@ -0,0 +1,74 @@
+    """Base class for a cog."""
+
+    def __init_subclass__(cls, **kwargs) -> None:
+        cls.__translator = kwargs.pop("translator", lambda s, t: t)
```
This should default to `None` and compare to `None` where needed. None type comparison is significantly cheaper than a lambda when considering every user-facing string.

Tobotimus committed:
Make `__translator` default to None

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

mikeshardmind commented on redbot/core/commands/cog.py:
```diff
@@ -0,0 +1,75 @@
+            return cls.__translator(untranslated)
+
+
+class Cog(metaclass=CogMeta):
```
Edit: the below changes are un-needed. It's been a rarity I've needed multiple metaclasses at once, so I had some alarm bells going off from a slightly different, but importantly syntactically different, case.
____
<details><summary>outdated</summary><p>
I'd prefer if the stuff in this were one more level of abstracted to better allow metaclass compatability.

This would entail chainging this to:

```py
class CogBase:
    """Base class for a cog."""
    def __init_subclass__(cls, **kwargs) -> None:
        cls.__translator = kwargs.pop("translator", None)
        super().__init_subclass__(**kwargs)

    @classmethod
    async def convert(cls, ctx: "Context", argument: str) -> "Cog":
        cog_instance = ctx.bot.get_cog(argument)
        if cog_instance is None:
            raise discord.ext.commands.BadArgument(_('Cog "argument" not found.'))
        return cog_instance

class Cog(CogBase, metaclass=CogMeta):
    pass
```

Doing this could allow for cooperative metaclassing in the following manner:

```py
# .abcs
from redbot.core import commands

class _MyMeta(type):
    ...

class MyMeta(_MyMeta, commands.CogMeta):
    pass

class CogBase(commands.CogBase, metaclass=MyMeta):
    pass
```
While I understand this is an edge case, it's reasonably accomodated, and avoids things like nearly completely breaking cogs using metaclasses to define APIs for static typing tools like mypy (see python/mypy#1996 )
</p></details>

it should be enough for anyone needing this to do
```py
# .abcs
from redbot.core import commands

class MyMeta(commands.CogMeta):
    pass

class MyCogBase(commands.Cog, metaclass=MyMeta):
    pass
```

Tobotimus committed:
Make `checks` a private attribute

This is to make the attribute less likely to clash with attributes of cog instances. The `checks` attribute will instead only be accessible through `type(cog).checks`.

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Tobotimus committed:
Remove unexpected arg from redbot.core.errors translator

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Tobotimus committed:
KILL THE KOALA

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Tobotimus committed:
Move __checks assignment

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Tobotimus committed:
Fix __translator assignment

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Tobotimus committed:
Reformat

Signed-off-by: Toby Harradine <tobyharradine@gmail.com>

Tobotimus commented:
Currently rewriting this, will open a new PR when it's ready
